runtime-light/coroutine/shared-task.h

```diff
@@ -0,0 +1,361 @@
+    m_state = state::end;
+  }
+
+  constexpr bool resumable() const noexcept {
+    return m_coro.promise().ready();
+  }
+
+  constexpr void cancel() noexcept {
+    m_state = state::end;
+    m_coro.promise().cancel_awaiter(m_waiter);
+  }
+};
+
+} // namespace shared_task_impl_
+
+template<typename T>
+struct shared_task_t final {
+  template<std::same_as<T> F>
+  struct promise_non_void_t;
+  struct promise_void_t;
+
+  using promise_type = std::conditional_t<!std::is_void_v<T>, promise_non_void_t<T>, promise_void_t>;
+
+  constexpr explicit shared_task_t(std::coroutine_handle<> coro) noexcept
+    : m_haddress(coro.address()) {}
+
+  constexpr shared_task_t(const shared_task_t &other) noexcept
+    : m_haddress(other.m_haddress) {
+    if (m_haddress != nullptr) [[likely]] {
+      std::coroutine_handle<promise_type>::from_address(m_haddress).promise().add_ref();
+    }
+  }
+
+  constexpr shared_task_t(shared_task_t &&other) noexcept
+    : m_haddress(std::exchange(other.m_haddress, nullptr)) {}
+
+  shared_task_t &operator=(const shared_task_t &other) noexcept {
+    if (m_haddress != other.m_haddress) [[likely]] {
+      destroy();
+      m_haddress = other.m_haddress;
+      if (m_haddress) [[likely]] {
+        std::coroutine_handle<promise_type>::from_address(m_haddress).promise().add_ref();
+      }
+    }
+    return *this;
+  }
+
+  shared_task_t &operator=(shared_task_t &&other) noexcept {
+    if (this != std::addressof(other)) [[likely]] {
+      destroy();
+      m_haddress = std::exchange(other.m_haddress, nullptr);
+    }
+    return *this;
+  }
+
+  ~shared_task_t() {
+    if (m_haddress) {
+      destroy();
+    }
+  }
+
+  struct promise_base_t : public shared_task_impl_::promise_base_t<promise_type> {
+    constexpr shared_task_t get_return_object() noexcept {
+      return shared_task_t{std::coroutine_handle<promise_type>::from_promise(*static_cast<promise_type *>(this))};
+    }
+
+    static shared_task_t get_return_object_on_allocation_failure() noexcept {
+      php_critical_error("cannot allocate memory for shared_task_t");
+    }
+  };
+
+  template<std::same_as<T> F>
+  struct promise_non_void_t final : public promise_base_t {
+    constexpr promise_non_void_t() noexcept = default;
+    promise_non_void_t(const promise_non_void_t &other) = delete;
+    promise_non_void_t(promise_non_void_t &&other) = delete;
+    promise_non_void_t &operator=(const promise_non_void_t &other) = delete;
+    promise_non_void_t &operator=(promise_non_void_t &&other) = delete;
+
+    ~promise_non_void_t() {
+      std::launder(reinterpret_cast<T *>(m_bytes))->~T();
+    }
+
+    template<typename E>
+    requires std::constructible_from<F, E &&> constexpr void return_value(E &&e) noexcept {
+      ::new (m_bytes) F(std::forward<E>(e));
+    }
+
+    constexpr const T &result() const noexcept {
+      return *std::launder(reinterpret_cast<const T *>(m_bytes));
+    }
+
+    alignas(F) std::byte m_bytes[sizeof(F)]{};
```

Comment:
Why not `std::array<std::byte>`?

Reply:
I use C-style array here to reduce the complexity